mungegithub/mungers/publish_scripts/util.sh

```diff
@@ -19,118 +19,555 @@
-       local cleanGodepJsonCommit=${previousBranchSHA}
-       git checkout ${cleanGodepJsonCommit} Godeps/Godeps.json
-       if git diff --cached --exit-code &>/dev/null; then
-           echo "no need to reset Godeps.json!"
-       else
-           git -c user.name="Kubernetes Publisher" -c user.email="k8s-publish-robot@users.noreply.github.com" commit -m "sync: reset Godeps.json" -- Godeps/Godeps.json
-       fi
-    fi
-
-    echo "commits to be cherry-picked:"
-    echo "${commits}"
-    echo ""
-    
-    while read commitSHA; do
-        if [[ -z "${commitSHA}" ]]; then
-            continue
+
+        # is it a merge or a single commit on the mainline to apply?
+        if [ ${dst_branch} != master ] && is-merge-with-master ${k_mainline_commit}; then
+            echo "Deferring master merge commit ${k_mainline_commit}: $(git show --format="%s" -q ${f_mainline_commit})."
+        elif [ ${dst_branch} != master ] && [ -n "${k_pending_merge_commit}" ] && is-merge-with-master "${k_pending_merge_commit}"; then
+            echo "Skipping master commit ${k_mainline_commit}: $(git show --format="%s" -q ${f_mainline_commit}). Master merge commit ${k_pending_merge_commit} is pending."
+        elif ! git show -q ${f_mainline_commit} | grep -q "^Merge: "; then
+            echo "Cherry-picking k8s.io/kubernetes single-commit ${k_mainline_commit}: $(git show --format="%s" -q ${f_mainline_commit})."
+
+            # reset Godeps.json?
+            local squash_commits=1
+            if [ -n "$(godep-changes ${f_mainline_commit})" ]; then
+                reset-godeps ${f_mainline_commit}^
+                squash_commits=2 # squash the cherry-pick into the godep reset commit below
+                dst_needs_godeps_update=true
+            fi
+
+            # finally cherry-pick
+            local date=$(git show --format="%aD" -q ${f_mainline_commit})
+            if ! GIT_COMMITTER_DATE="${date}" git cherry-pick --keep-redundant-commits ${f_mainline_commit} >/dev/null; then
+                echo
+                show-working-dir-status
+                return 1
+            fi
+
+            # potentially squash godep reset commit
+            squash ${squash_commits}
+
+            # if there is no pending merge commit, update Godeps.json because this could be a target of tag
+            if [ -z "${k_pending_merge_commit}" ]; then
+                fix-godeps "${deps}" "${is_library}" ${dst_needs_godeps_update}
+                dst_needs_godeps_update=false
+                dst_merge_point_commit=$(git rev-parse HEAD)
+            fi
+        else
+            # find **latest** (in the sense of least distance to ${f_mainline_commit}) common ancestor of both parents
+            # of ${f_mainline_commit}. If the PR had no merge commits, this is the actual fork point. If there have been
+            # merges, we will get the latest of those merges. Everything between that and ${f_mainline_commit} will be
```

Comment:
> we will get the latest of those merges

It should be the a commit in the mainline of kubernetes, that is the latest ancestor of the last "merges in the PR", right?

The algorithm looks right, the comment is misleading.

Reply:
> "the common ancestor" will be the a commit in the mainline of kubernetes

Not necessarily. I have added two pictures. A picture is worth a thousand words....